File: apps/query-burst-msa/catalog-service/src/main/java/com/booster/queryburstmsa/catalog/domain/repository/ProductRepository.java
package com.booster.queryburstmsa.catalog.domain.repository;

import com.booster.queryburstmsa.catalog.domain.ProductStatus;
import com.booster.queryburstmsa.catalog.domain.entity.ProductEntity;
import jakarta.persistence.LockModeType;
import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Lock;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;

import java.util.List;

public interface ProductRepository extends JpaRepository<ProductEntity, Long> {

    @Query("""
            select p
            from ProductEntity p
            where (:cursor is null or p.id < :cursor)
              and (:categoryId is null or p.categoryId = :categoryId)
              and (:status is null or p.status = :status)
            order by p.id desc
            """)
    List<ProductEntity> findProducts(
            @Param("cursor") Long cursor,
            @Param("categoryId") Long categoryId,
            @Param("status") ProductStatus status,
            Pageable pageable
    );

    @Lock(LockModeType.PESSIMISTIC_WRITE)
    @Query("select p from ProductEntity p where p.id in :ids")
    List<ProductEntity> findAllByIdInForUpdate(@Param("ids") List<Long> ids);
}

File: apps/kotlin/shopping-service/src/main/kotlin/com/booster/kotlin/shoppingservice/catalog/domain/ProductRepository.kt
package com.booster.kotlin.shoppingservice.catalog.domain

import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Query
import org.springframework.data.repository.query.Param

interface ProductRepository : JpaRepository<Product, Long>{

    @Query(
        """
        SELECT p FROM Product p
        WHERE p.status = :status
        AND (:categoryId IS NULL OR p.category.id = :categoryId)
    """
    )
    fun findOnSaleProducts(
        @Param("status") status: ProductStatus,
        @Param("categoryId") categoryId: Long?,
        pageable: Pageable,
    ): Page<Product>
}